{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "cd7d910a",
   "metadata": {},
   "source": [
    "1. Dataset Analysis: Explore the statistics and potential issues of the dataset.\n",
    "2. Method Explanation: You should explain the method(s) you used.\n",
    "3. Experiment Description: You should design a set of experiments in order to prove the advantage (if exists) of the proposed method.\n",
    "4. Result Analysis: You should resume your experiment results and analyze them.\n",
    "\n",
    "\n",
    "Multiple users in Kaggle have already conducted comprehensive analyses of the PFAM database. In this study, we leverage some of the existing tools to hilight certain aspects which have not be fully discussed yet. Bileschi et al. in their paper titled \"Using Deep Learning to Annotate the Protein Universe\" they present a remarkable convolutional neural network (CNN) surpasses the accuracy of both BLAST, which operates through pairwise alignment, and hidden Markov models (pHMMs) based on scoring functions. This demonstrates the superior performance of the CNN in classifying the protein universe. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "091a2a8f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "\n",
    "import tensorflow \n",
    "import numpy \n",
    "import pandas\n",
    "import matplotlib.pyplot as plt\n",
    "from mpl_toolkits.axes_grid1.inset_locator import inset_axes\n",
    "\n",
    "import seaborn\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "239c2b64",
   "metadata": {},
   "source": [
    "## Dataset Analysis:\n",
    "Start by loading and exploring the dataset from the Pfam seed random split. Analyze the statistics of the dataset, such as the number of protein samples, the distribution of Pfam families, and any class imbalance issues. We visualize the data to gain insights, such as plotting the distribution of protein sequence lengths, examining the relationship between different features!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "8ce02173",
   "metadata": {},
   "outputs": [],
   "source": [
    "split_dataset = '/home/or22568/Desktop/PFAM/random_split' \n",
    "train_dataset_folder = f'{split_dataset}/train/'\n",
    "test_dataset_folder = f'{split_dataset}/test/'\n",
    "dev_dataset_folder = f'{split_dataset}/dev/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "499a9dd4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# We use this help function to read the datasets\n",
    "def read_data_csv(data_set_folder: str) -> pandas.DataFrame:\n",
    "    files = []\n",
    "    for file_name in os.scandir(data_set_folder): # scandir to get all files in the folder\n",
    "        files.append(pandas.read_csv(file_name))  # array of pandas.DataFrame\n",
    "    csv_pandas = pandas.concat(files)             # Concatenate the pandas.DataFrames\n",
    "    return csv_pandas "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "64a8aac4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# We read the train, test and dev datasets present in the three folders \n",
    "train_dataset = read_data_csv(train_dataset_folder)\n",
    "test_dataset = read_data_csv(test_dataset_folder)\n",
    "dev_dataset = read_data_csv(dev_dataset_folder)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "60583559",
   "metadata": {},
   "source": [
    "Data Structure (see https://www.kaggle.com/datasets/googleai/pfam-seed-random-split?resource=download): \n",
    "\n",
    "Each fold (train, dev, test) has a number of files in it. Each of those files\n",
    "contains csv on each line, which has the following fields:\n",
    "\n",
    "### sequence\n",
    "### family_accession\n",
    "### sequence_name\n",
    "### aligned_sequence\n",
    "### family_id\n",
    "\n",
    "Description of fields:\n",
    "\n",
    "sequence: These are usually the input features to your model. Amino acid sequence for this domain.\n",
    "There are 20 very common amino acids (frequency > 1,000,000), and 4 amino acids that are quite\n",
    "uncommon: X, U, B, O, Z. This can be the training feature! \n",
    "\n",
    "family_accession: These are usually the labels for your model. Accession number in form PFxxxxx.y\n",
    "(Pfam), where xxxxx is the family accession, and y is the version number. This is usually the Label!\n",
    "\n",
    "family_id: One word name for family.\n",
    "\n",
    "sequence_name: Sequence name, in the form \"uniprot accession id/start index // end_index\".\n",
    "\n",
    "aligned_sequence: Contains a single sequence from the multiple sequence alignment (with the rest of the members of\n",
    "the family in seed, with gaps retained. This can be the training feature! "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "60a8d3b4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>family_id</th>\n",
       "      <th>sequence_name</th>\n",
       "      <th>family_accession</th>\n",
       "      <th>aligned_sequence</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GNAT_acetyltran</td>\n",
       "      <td>R6RQF6_9CLOT/17-251</td>\n",
       "      <td>PF12746.7</td>\n",
       "      <td>AFLFSGR..REVMAD....ACLQGMM..GCVYG..........TAG...</td>\n",
       "      <td>AFLFSGRREVMADACLQGMMGCVYGTAGGMDSAAAVLGDFCFLAGK...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MoaC</td>\n",
       "      <td>W5NKR5_LEPOC/505-640</td>\n",
       "      <td>PF01967.21</td>\n",
       "      <td>MVDVGGK.PVSRRTAAASATVLLG.EK..........AFWLV.......</td>\n",
       "      <td>MVDVGGKPVSRRTAAASATVLLGEKAFWLVKENQLAKGDALAVAQI...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Methyltransf_25</td>\n",
       "      <td>C0QLU8_DESAH/50-147</td>\n",
       "      <td>PF13649.6</td>\n",
       "      <td>VLDVACGT.C...D..VA...ME..AR.NQ.......T....G......</td>\n",
       "      <td>VLDVACGTCDVAMEARNQTGDAAFIIGTDFSPGMLTLGLQKLKKNR...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>EMG1</td>\n",
       "      <td>T1G7Q2_HELRO/22-222</td>\n",
       "      <td>PF03587.14</td>\n",
       "      <td>VVLERASLESVKV..G.................KEYQLLN....CD...</td>\n",
       "      <td>VVLERASLESVKVGKEYQLLNCDRHKGIAKKFKRDISTCRPDITHQ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Glyco_hydro_30C</td>\n",
       "      <td>C6VRM9_DYAFD/453-540</td>\n",
       "      <td>PF17189.4</td>\n",
       "      <td>GAVRVDVSGGLGTD...............AMVVSSYLN..TDKSLV...</td>\n",
       "      <td>GAVRVDVSGGLGTDAMVVSSYLNTDKSLVTVIVNADNQDRDISLAI...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         family_id         sequence_name family_accession  \\\n",
       "0  GNAT_acetyltran   R6RQF6_9CLOT/17-251        PF12746.7   \n",
       "1             MoaC  W5NKR5_LEPOC/505-640       PF01967.21   \n",
       "2  Methyltransf_25   C0QLU8_DESAH/50-147        PF13649.6   \n",
       "3             EMG1   T1G7Q2_HELRO/22-222       PF03587.14   \n",
       "4  Glyco_hydro_30C  C6VRM9_DYAFD/453-540        PF17189.4   \n",
       "\n",
       "                                    aligned_sequence  \\\n",
       "0  AFLFSGR..REVMAD....ACLQGMM..GCVYG..........TAG...   \n",
       "1  MVDVGGK.PVSRRTAAASATVLLG.EK..........AFWLV.......   \n",
       "2  VLDVACGT.C...D..VA...ME..AR.NQ.......T....G......   \n",
       "3  VVLERASLESVKV..G.................KEYQLLN....CD...   \n",
       "4  GAVRVDVSGGLGTD...............AMVVSSYLN..TDKSLV...   \n",
       "\n",
       "                                            sequence  \n",
       "0  AFLFSGRREVMADACLQGMMGCVYGTAGGMDSAAAVLGDFCFLAGK...  \n",
       "1  MVDVGGKPVSRRTAAASATVLLGEKAFWLVKENQLAKGDALAVAQI...  \n",
       "2  VLDVACGTCDVAMEARNQTGDAAFIIGTDFSPGMLTLGLQKLKKNR...  \n",
       "3  VVLERASLESVKVGKEYQLLNCDRHKGIAKKFKRDISTCRPDITHQ...  \n",
       "4  GAVRVDVSGGLGTDAMVVSSYLNTDKSLVTVIVNADNQDRDISLAI...  "
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# We visually inspect the pandas.DataFrame initial fields\n",
    "test_dataset.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "0c7e3c42",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>family_id</th>\n",
       "      <th>sequence_name</th>\n",
       "      <th>family_accession</th>\n",
       "      <th>aligned_sequence</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>12751</th>\n",
       "      <td>C1-set</td>\n",
       "      <td>IGHG2_CAVPO/229-315</td>\n",
       "      <td>PF07654.15</td>\n",
       "      <td>VYTLPPSRDE.LS.KSKVSVTCLIINFFP..ADIHVEWASNRVPVS...</td>\n",
       "      <td>VYTLPPSRDELSKSKVSVTCLIINFFPADIHVEWASNRVPVSEKEY...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12752</th>\n",
       "      <td>Cofac_haem_bdg</td>\n",
       "      <td>F0LPX5_VIBFN/30-233</td>\n",
       "      <td>PF04187.13</td>\n",
       "      <td>FRHAIRHADVILVGEWH..THTGIHRFQTDLLQTL..............</td>\n",
       "      <td>FRHAIRHADVILVGEWHTHTGIHRFQTDLLQTLSQEERPVALSMEQ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12753</th>\n",
       "      <td>S-methyl_trans</td>\n",
       "      <td>A1B3D4_PARDP/21-337</td>\n",
       "      <td>PF02574.16</td>\n",
       "      <td>LI.....................LDGAMGTQI.Q....QL.........</td>\n",
       "      <td>LILDGAMGTQIQQLGLSETDFAGHGTGCACGCHPPAPGEHPQQGNN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12754</th>\n",
       "      <td>Glyco_trans_1_4</td>\n",
       "      <td>B8FJ84_DESAA/245-385</td>\n",
       "      <td>PF13692.6</td>\n",
       "      <td>KNRLLVTN...SA.DT.....PLKGL.YHL..LKAV.HEVR....K...</td>\n",
       "      <td>KNRLLVTNSADTPLKGLYHLLKAVHEVRKKRDVTLTVIGAPKKHGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12755</th>\n",
       "      <td>OB_MalK</td>\n",
       "      <td>A0A0A0D4M6_9PROT/236-288</td>\n",
       "      <td>PF17912.1</td>\n",
       "      <td>GSPAMNFLPGT..V.V..RSDG.......LVARIDGQDL.PL...G...</td>\n",
       "      <td>GSPAMNFLPGTVVRSDGLVARIDGQDLPLGAYPFLSEPAAGDEIVL...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             family_id             sequence_name family_accession  \\\n",
       "12751           C1-set       IGHG2_CAVPO/229-315       PF07654.15   \n",
       "12752   Cofac_haem_bdg       F0LPX5_VIBFN/30-233       PF04187.13   \n",
       "12753   S-methyl_trans       A1B3D4_PARDP/21-337       PF02574.16   \n",
       "12754  Glyco_trans_1_4      B8FJ84_DESAA/245-385        PF13692.6   \n",
       "12755          OB_MalK  A0A0A0D4M6_9PROT/236-288        PF17912.1   \n",
       "\n",
       "                                        aligned_sequence  \\\n",
       "12751  VYTLPPSRDE.LS.KSKVSVTCLIINFFP..ADIHVEWASNRVPVS...   \n",
       "12752  FRHAIRHADVILVGEWH..THTGIHRFQTDLLQTL..............   \n",
       "12753  LI.....................LDGAMGTQI.Q....QL.........   \n",
       "12754  KNRLLVTN...SA.DT.....PLKGL.YHL..LKAV.HEVR....K...   \n",
       "12755  GSPAMNFLPGT..V.V..RSDG.......LVARIDGQDL.PL...G...   \n",
       "\n",
       "                                                sequence  \n",
       "12751  VYTLPPSRDELSKSKVSVTCLIINFFPADIHVEWASNRVPVSEKEY...  \n",
       "12752  FRHAIRHADVILVGEWHTHTGIHRFQTDLLQTLSQEERPVALSMEQ...  \n",
       "12753  LILDGAMGTQIQQLGLSETDFAGHGTGCACGCHPPAPGEHPQQGNN...  \n",
       "12754  KNRLLVTNSADTPLKGLYHLLKAVHEVRKKRDVTLTVIGAPKKHGG...  \n",
       "12755  GSPAMNFLPGTVVRSDGLVARIDGQDLPLGAYPFLSEPAAGDEIVL...  "
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# We visually inspect the pandas.DataFrame final fields\n",
    "test_dataset.tail()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f1485c57",
   "metadata": {},
   "source": [
    "One crucial aspect to examine first is whether all the protein families are reported in the three subfolders and the number of proteins present in each subfolder."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "134ba143",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_partition_information(all_dataset: dict) -> pandas.DataFrame:\n",
    "    columns = ['dataset', 'num_samples', 'num_families', 'min_samples_per_family', 'min_family', 'max_samples_per_family', 'max_family', 'mean_samples_per_family', 'name']\n",
    "    data_analysis_info = pandas.DataFrame(columns=columns)\n",
    "    \n",
    "    for dataset_name, dataset in all_dataset.items():\n",
    "        num_samples = len(dataset) # Total number of points\n",
    "        num_families = dataset['family_accession'].nunique() # Number of nonunique family_accession\n",
    "        fam_sizes = dataset.groupby('family_accession').size() # Number of protein for each family\n",
    "        max_family = fam_sizes.idxmax() # Get the family corresponding to the maximum samples  \n",
    "        min_family = fam_sizes.idxmin() # Get the family corresponding to the minimum samples  \n",
    "        min_samples_per_family = fam_sizes.min()\n",
    "        max_samples_per_family = fam_sizes.max()\n",
    "        mean_samples_per_family = fam_sizes.mean()\n",
    "\n",
    "        data_analysis_info = pandas.concat([data_analysis_info, pandas.DataFrame({'dataset': dataset_name,\n",
    "                                                                                  'num_samples': num_samples,\n",
    "                                                                                  'num_families': num_families,\n",
    "                                                                                  'min_samples_per_family': min_samples_per_family,\n",
    "                                                                                  'max_samples_per_family': max_samples_per_family,\n",
    "                                                                                  'mean_samples_per_family': mean_samples_per_family,\n",
    "                                                                                  'name': dataset_name,\n",
    "                                                                                  'min_family': min_family,\n",
    "                                                                                  'max_family': max_family}, index=[0])],\n",
    "                                           ignore_index=True)\n",
    "    return data_analysis_info\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "5bc14fc1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dataset</th>\n",
       "      <th>num_samples</th>\n",
       "      <th>num_families</th>\n",
       "      <th>min_samples_per_family</th>\n",
       "      <th>min_family</th>\n",
       "      <th>max_samples_per_family</th>\n",
       "      <th>max_family</th>\n",
       "      <th>mean_samples_per_family</th>\n",
       "      <th>name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>train</td>\n",
       "      <td>1086741</td>\n",
       "      <td>17929</td>\n",
       "      <td>1</td>\n",
       "      <td>PF00541.17</td>\n",
       "      <td>3637</td>\n",
       "      <td>PF13649.6</td>\n",
       "      <td>60.613587</td>\n",
       "      <td>train</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>test</td>\n",
       "      <td>126171</td>\n",
       "      <td>13071</td>\n",
       "      <td>1</td>\n",
       "      <td>PF00058.17</td>\n",
       "      <td>454</td>\n",
       "      <td>PF13649.6</td>\n",
       "      <td>9.652743</td>\n",
       "      <td>test</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>dev</td>\n",
       "      <td>126171</td>\n",
       "      <td>13071</td>\n",
       "      <td>1</td>\n",
       "      <td>PF00058.17</td>\n",
       "      <td>454</td>\n",
       "      <td>PF13649.6</td>\n",
       "      <td>9.652743</td>\n",
       "      <td>dev</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  dataset num_samples num_families min_samples_per_family  min_family  \\\n",
       "0   train     1086741        17929                      1  PF00541.17   \n",
       "1    test      126171        13071                      1  PF00058.17   \n",
       "2     dev      126171        13071                      1  PF00058.17   \n",
       "\n",
       "  max_samples_per_family max_family  mean_samples_per_family   name  \n",
       "0                   3637  PF13649.6                60.613587  train  \n",
       "1                    454  PF13649.6                 9.652743   test  \n",
       "2                    454  PF13649.6                 9.652743    dev  "
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_dataset = {'train': train_dataset, 'test': test_dataset, 'dev': dev_dataset}\n",
    "get_partition_information(all_dataset)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a1d2094f",
   "metadata": {},
   "source": [
    "Interestingly enough the protein family with less instances (1) in the \"test\" and \"dev\" databases is not the same of the train databse! In the following cell we notice that there are 17 instances of PF00058.17 in the train set.\n",
    "\n",
    "PF13649.6 is instead the family more rapresented in all datasets. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "075ee3e5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The number of instances of protein with family_accession PF00058.17: 17\n"
     ]
    }
   ],
   "source": [
    "def get_protein_count(dataset: pandas.DataFrame, family_accession: str) -> int:\n",
    "    count = len(dataset[dataset['family_accession'] == family_accession])\n",
    "    return count\n",
    "\n",
    "family_accession = 'PF00058.17'\n",
    "count = get_protein_count(train_dataset, family_accession)\n",
    "print(f\"The number of instances of protein with family_accession {family_accession}: {count}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "64c3449d",
   "metadata": {},
   "source": [
    "Upon initial analysis, it is evident that certain protein families are exclusively present in the training data, as indicated by the considerably higher number of families in the train partition compared to the other partitions. However, it remains uncertain whether there are families exclusive to the dev or test data. To gain further insights, we will delve deeper into the dataset and investigate if any families are exclusive to these partitions as well."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "6f32d521",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "dev and test familes are the same\n"
     ]
    }
   ],
   "source": [
    "def intersection_common_families(dataset: pandas.DataFrame, dataset2: pandas.DataFrame) -> pandas.DataFrame: \n",
    "    return dataset.intersection(dataset2)\n",
    "\n",
    "\n",
    "train_families = set(train_dataset['family_accession'])\n",
    "dev_families = set(dev_dataset['family_accession'])\n",
    "test_families = set(test_dataset['family_accession'])\n",
    "\n",
    "if dev_families == test_families: \n",
    "    print(f'dev and test familes are the same')\n",
    "    common_families = intersection_common_families(train_families, dev_families)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5ebcc517",
   "metadata": {},
   "source": [
    "Based on the analysis, it is observed that there are exclusively unique protein families present only in the training dataset. To ensure a consistent validation process, it is advisable to exclude these families from the analysis. By doing so, the number of families will be consistent across all datasets (train, dev, and test). This approach ensures that the mapping of families can be effectively validated using the dev and test sets without any discrepancies."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "9c73f827",
   "metadata": {},
   "outputs": [],
   "source": [
    "train_dataset = train_dataset[train_dataset['family_accession'].isin(common_families)]\n",
    "all_dataset['train'] = train_dataset"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2b6e774d",
   "metadata": {},
   "source": [
    "In the paper of Bileschi et al. they noticed that some of the protein were much less represented than the others. We explore here some of the underlined statistics "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "id": "367b8826",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_227015/3561229292.py:25: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.\n",
      "  plt.tight_layout()\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[90], line 26\u001b[0m\n\u001b[1;32m     23\u001b[0m     ax_inset\u001b[38;5;241m.\u001b[39mset_ylabel(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mNumber of Families\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m     25\u001b[0m plt\u001b[38;5;241m.\u001b[39mtight_layout()\n\u001b[0;32m---> 26\u001b[0m \u001b[43mplt\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mshow\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/envs/bio-compchem/lib/python3.8/site-packages/matplotlib/pyplot.py:446\u001b[0m, in \u001b[0;36mshow\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m    402\u001b[0m \u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m    403\u001b[0m \u001b[38;5;124;03mDisplay all open figures.\u001b[39;00m\n\u001b[1;32m    404\u001b[0m \n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    443\u001b[0m \u001b[38;5;124;03mexplicitly there.\u001b[39;00m\n\u001b[1;32m    444\u001b[0m \u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m    445\u001b[0m _warn_if_gui_out_of_main_thread()\n\u001b[0;32m--> 446\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43m_get_backend_mod\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mshow\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/envs/bio-compchem/lib/python3.8/site-packages/matplotlib_inline/backend_inline.py:90\u001b[0m, in \u001b[0;36mshow\u001b[0;34m(close, block)\u001b[0m\n\u001b[1;32m     88\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m     89\u001b[0m     \u001b[38;5;28;01mfor\u001b[39;00m figure_manager \u001b[38;5;129;01min\u001b[39;00m Gcf\u001b[38;5;241m.\u001b[39mget_all_fig_managers():\n\u001b[0;32m---> 90\u001b[0m         \u001b[43mdisplay\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m     91\u001b[0m \u001b[43m            \u001b[49m\u001b[43mfigure_manager\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcanvas\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mfigure\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     92\u001b[0m \u001b[43m            \u001b[49m\u001b[43mmetadata\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m_fetch_figure_metadata\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfigure_manager\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcanvas\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mfigure\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     93\u001b[0m \u001b[43m        \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     94\u001b[0m \u001b[38;5;28;01mfinally\u001b[39;00m:\n\u001b[1;32m     95\u001b[0m     show\u001b[38;5;241m.\u001b[39m_to_draw \u001b[38;5;241m=\u001b[39m []\n",
      "File \u001b[0;32m~/miniconda3/envs/bio-compchem/lib/python3.8/site-packages/IPython/core/display_functions.py:298\u001b[0m, in \u001b[0;36mdisplay\u001b[0;34m(include, exclude, metadata, transient, display_id, raw, clear, *objs, **kwargs)\u001b[0m\n\u001b[1;32m    296\u001b[0m     publish_display_data(data\u001b[38;5;241m=\u001b[39mobj, metadata\u001b[38;5;241m=\u001b[39mmetadata, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs)\n\u001b[1;32m    297\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m--> 298\u001b[0m     format_dict, md_dict \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mformat\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43mobj\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43minclude\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43minclude\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mexclude\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mexclude\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    299\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m format_dict:\n\u001b[1;32m    300\u001b[0m         \u001b[38;5;66;03m# nothing to display (e.g. _ipython_display_ took over)\u001b[39;00m\n\u001b[1;32m    301\u001b[0m         \u001b[38;5;28;01mcontinue\u001b[39;00m\n",
      "File \u001b[0;32m~/miniconda3/envs/bio-compchem/lib/python3.8/site-packages/IPython/core/formatters.py:179\u001b[0m, in \u001b[0;36mDisplayFormatter.format\u001b[0;34m(self, obj, include, exclude)\u001b[0m\n\u001b[1;32m    177\u001b[0m md \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[1;32m    178\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 179\u001b[0m     data \u001b[38;5;241m=\u001b[39m \u001b[43mformatter\u001b[49m\u001b[43m(\u001b[49m\u001b[43mobj\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    180\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m:\n\u001b[1;32m    181\u001b[0m     \u001b[38;5;66;03m# FIXME: log the exception\u001b[39;00m\n\u001b[1;32m    182\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m\n",
      "File \u001b[0;32m~/miniconda3/envs/bio-compchem/lib/python3.8/site-packages/decorator.py:232\u001b[0m, in \u001b[0;36mdecorate.<locals>.fun\u001b[0;34m(*args, **kw)\u001b[0m\n\u001b[1;32m    230\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m kwsyntax:\n\u001b[1;32m    231\u001b[0m     args, kw \u001b[38;5;241m=\u001b[39m fix(args, kw, sig)\n\u001b[0;32m--> 232\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mcaller\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfunc\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43mextras\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m+\u001b[39;49m\u001b[43m \u001b[49m\u001b[43margs\u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkw\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/envs/bio-compchem/lib/python3.8/site-packages/IPython/core/formatters.py:223\u001b[0m, in \u001b[0;36mcatch_format_error\u001b[0;34m(method, self, *args, **kwargs)\u001b[0m\n\u001b[1;32m    221\u001b[0m \u001b[38;5;124;03m\"\"\"show traceback on failed format call\"\"\"\u001b[39;00m\n\u001b[1;32m    222\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 223\u001b[0m     r \u001b[38;5;241m=\u001b[39m \u001b[43mmethod\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    224\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mNotImplementedError\u001b[39;00m:\n\u001b[1;32m    225\u001b[0m     \u001b[38;5;66;03m# don't warn on NotImplementedErrors\u001b[39;00m\n\u001b[1;32m    226\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_check_return(\u001b[38;5;28;01mNone\u001b[39;00m, args[\u001b[38;5;241m0\u001b[39m])\n",
      "File \u001b[0;32m~/miniconda3/envs/bio-compchem/lib/python3.8/site-packages/IPython/core/formatters.py:340\u001b[0m, in \u001b[0;36mBaseFormatter.__call__\u001b[0;34m(self, obj)\u001b[0m\n\u001b[1;32m    338\u001b[0m     \u001b[38;5;28;01mpass\u001b[39;00m\n\u001b[1;32m    339\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m--> 340\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mprinter\u001b[49m\u001b[43m(\u001b[49m\u001b[43mobj\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    341\u001b[0m \u001b[38;5;66;03m# Finally look for special method names\u001b[39;00m\n\u001b[1;32m    342\u001b[0m method \u001b[38;5;241m=\u001b[39m get_real_method(obj, \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mprint_method)\n",
      "File \u001b[0;32m~/miniconda3/envs/bio-compchem/lib/python3.8/site-packages/IPython/core/pylabtools.py:152\u001b[0m, in \u001b[0;36mprint_figure\u001b[0;34m(fig, fmt, bbox_inches, base64, **kwargs)\u001b[0m\n\u001b[1;32m    149\u001b[0m     \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mmatplotlib\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mbackend_bases\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m FigureCanvasBase\n\u001b[1;32m    150\u001b[0m     FigureCanvasBase(fig)\n\u001b[0;32m--> 152\u001b[0m \u001b[43mfig\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcanvas\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mprint_figure\u001b[49m\u001b[43m(\u001b[49m\u001b[43mbytes_io\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkw\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    153\u001b[0m data \u001b[38;5;241m=\u001b[39m bytes_io\u001b[38;5;241m.\u001b[39mgetvalue()\n\u001b[1;32m    154\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m fmt \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124msvg\u001b[39m\u001b[38;5;124m'\u001b[39m:\n",
      "File \u001b[0;32m~/miniconda3/envs/bio-compchem/lib/python3.8/site-packages/matplotlib/backend_bases.py:2342\u001b[0m, in \u001b[0;36mFigureCanvasBase.print_figure\u001b[0;34m(self, filename, dpi, facecolor, edgecolor, orientation, format, bbox_inches, pad_inches, bbox_extra_artists, backend, **kwargs)\u001b[0m\n\u001b[1;32m   2336\u001b[0m     renderer \u001b[38;5;241m=\u001b[39m _get_renderer(\n\u001b[1;32m   2337\u001b[0m         \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mfigure,\n\u001b[1;32m   2338\u001b[0m         functools\u001b[38;5;241m.\u001b[39mpartial(\n\u001b[1;32m   2339\u001b[0m             print_method, orientation\u001b[38;5;241m=\u001b[39morientation)\n\u001b[1;32m   2340\u001b[0m     )\n\u001b[1;32m   2341\u001b[0m     \u001b[38;5;28;01mwith\u001b[39;00m \u001b[38;5;28mgetattr\u001b[39m(renderer, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m_draw_disabled\u001b[39m\u001b[38;5;124m\"\u001b[39m, nullcontext)():\n\u001b[0;32m-> 2342\u001b[0m         \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mfigure\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mdraw\u001b[49m\u001b[43m(\u001b[49m\u001b[43mrenderer\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   2344\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m bbox_inches:\n\u001b[1;32m   2345\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m bbox_inches \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mtight\u001b[39m\u001b[38;5;124m\"\u001b[39m:\n",
      "File \u001b[0;32m~/miniconda3/envs/bio-compchem/lib/python3.8/site-packages/matplotlib/artist.py:95\u001b[0m, in \u001b[0;36m_finalize_rasterization.<locals>.draw_wrapper\u001b[0;34m(artist, renderer, *args, **kwargs)\u001b[0m\n\u001b[1;32m     93\u001b[0m \u001b[38;5;129m@wraps\u001b[39m(draw)\n\u001b[1;32m     94\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mdraw_wrapper\u001b[39m(artist, renderer, \u001b[38;5;241m*\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs):\n\u001b[0;32m---> 95\u001b[0m     result \u001b[38;5;241m=\u001b[39m \u001b[43mdraw\u001b[49m\u001b[43m(\u001b[49m\u001b[43martist\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mrenderer\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     96\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m renderer\u001b[38;5;241m.\u001b[39m_rasterizing:\n\u001b[1;32m     97\u001b[0m         renderer\u001b[38;5;241m.\u001b[39mstop_rasterizing()\n",
      "File \u001b[0;32m~/miniconda3/envs/bio-compchem/lib/python3.8/site-packages/matplotlib/artist.py:72\u001b[0m, in \u001b[0;36mallow_rasterization.<locals>.draw_wrapper\u001b[0;34m(artist, renderer)\u001b[0m\n\u001b[1;32m     69\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m artist\u001b[38;5;241m.\u001b[39mget_agg_filter() \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m     70\u001b[0m         renderer\u001b[38;5;241m.\u001b[39mstart_filter()\n\u001b[0;32m---> 72\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mdraw\u001b[49m\u001b[43m(\u001b[49m\u001b[43martist\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mrenderer\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     73\u001b[0m \u001b[38;5;28;01mfinally\u001b[39;00m:\n\u001b[1;32m     74\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m artist\u001b[38;5;241m.\u001b[39mget_agg_filter() \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n",
      "File \u001b[0;32m~/miniconda3/envs/bio-compchem/lib/python3.8/site-packages/matplotlib/figure.py:3140\u001b[0m, in \u001b[0;36mFigure.draw\u001b[0;34m(self, renderer)\u001b[0m\n\u001b[1;32m   3137\u001b[0m         \u001b[38;5;66;03m# ValueError can occur when resizing a window.\u001b[39;00m\n\u001b[1;32m   3139\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mpatch\u001b[38;5;241m.\u001b[39mdraw(renderer)\n\u001b[0;32m-> 3140\u001b[0m \u001b[43mmimage\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_draw_list_compositing_images\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   3141\u001b[0m \u001b[43m    \u001b[49m\u001b[43mrenderer\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43martists\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msuppressComposite\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   3143\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m sfig \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39msubfigs:\n\u001b[1;32m   3144\u001b[0m     sfig\u001b[38;5;241m.\u001b[39mdraw(renderer)\n",
      "File \u001b[0;32m~/miniconda3/envs/bio-compchem/lib/python3.8/site-packages/matplotlib/image.py:131\u001b[0m, in \u001b[0;36m_draw_list_compositing_images\u001b[0;34m(renderer, parent, artists, suppress_composite)\u001b[0m\n\u001b[1;32m    129\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m not_composite \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m has_images:\n\u001b[1;32m    130\u001b[0m     \u001b[38;5;28;01mfor\u001b[39;00m a \u001b[38;5;129;01min\u001b[39;00m artists:\n\u001b[0;32m--> 131\u001b[0m         \u001b[43ma\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mdraw\u001b[49m\u001b[43m(\u001b[49m\u001b[43mrenderer\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    132\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    133\u001b[0m     \u001b[38;5;66;03m# Composite any adjacent images together\u001b[39;00m\n\u001b[1;32m    134\u001b[0m     image_group \u001b[38;5;241m=\u001b[39m []\n",
      "File \u001b[0;32m~/miniconda3/envs/bio-compchem/lib/python3.8/site-packages/matplotlib/artist.py:72\u001b[0m, in \u001b[0;36mallow_rasterization.<locals>.draw_wrapper\u001b[0;34m(artist, renderer)\u001b[0m\n\u001b[1;32m     69\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m artist\u001b[38;5;241m.\u001b[39mget_agg_filter() \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m     70\u001b[0m         renderer\u001b[38;5;241m.\u001b[39mstart_filter()\n\u001b[0;32m---> 72\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mdraw\u001b[49m\u001b[43m(\u001b[49m\u001b[43martist\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mrenderer\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     73\u001b[0m \u001b[38;5;28;01mfinally\u001b[39;00m:\n\u001b[1;32m     74\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m artist\u001b[38;5;241m.\u001b[39mget_agg_filter() \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n",
      "File \u001b[0;32m~/miniconda3/envs/bio-compchem/lib/python3.8/site-packages/matplotlib/axes/_base.py:3064\u001b[0m, in \u001b[0;36m_AxesBase.draw\u001b[0;34m(self, renderer)\u001b[0m\n\u001b[1;32m   3061\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m artists_rasterized:\n\u001b[1;32m   3062\u001b[0m     _draw_rasterized(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mfigure, artists_rasterized, renderer)\n\u001b[0;32m-> 3064\u001b[0m \u001b[43mmimage\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_draw_list_compositing_images\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   3065\u001b[0m \u001b[43m    \u001b[49m\u001b[43mrenderer\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43martists\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mfigure\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msuppressComposite\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   3067\u001b[0m renderer\u001b[38;5;241m.\u001b[39mclose_group(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124maxes\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[1;32m   3068\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mstale \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mFalse\u001b[39;00m\n",
      "File \u001b[0;32m~/miniconda3/envs/bio-compchem/lib/python3.8/site-packages/matplotlib/image.py:131\u001b[0m, in \u001b[0;36m_draw_list_compositing_images\u001b[0;34m(renderer, parent, artists, suppress_composite)\u001b[0m\n\u001b[1;32m    129\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m not_composite \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m has_images:\n\u001b[1;32m    130\u001b[0m     \u001b[38;5;28;01mfor\u001b[39;00m a \u001b[38;5;129;01min\u001b[39;00m artists:\n\u001b[0;32m--> 131\u001b[0m         \u001b[43ma\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mdraw\u001b[49m\u001b[43m(\u001b[49m\u001b[43mrenderer\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    132\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    133\u001b[0m     \u001b[38;5;66;03m# Composite any adjacent images together\u001b[39;00m\n\u001b[1;32m    134\u001b[0m     image_group \u001b[38;5;241m=\u001b[39m []\n",
      "File \u001b[0;32m~/miniconda3/envs/bio-compchem/lib/python3.8/site-packages/matplotlib/artist.py:72\u001b[0m, in \u001b[0;36mallow_rasterization.<locals>.draw_wrapper\u001b[0;34m(artist, renderer)\u001b[0m\n\u001b[1;32m     69\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m artist\u001b[38;5;241m.\u001b[39mget_agg_filter() \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m     70\u001b[0m         renderer\u001b[38;5;241m.\u001b[39mstart_filter()\n\u001b[0;32m---> 72\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mdraw\u001b[49m\u001b[43m(\u001b[49m\u001b[43martist\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mrenderer\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     73\u001b[0m \u001b[38;5;28;01mfinally\u001b[39;00m:\n\u001b[1;32m     74\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m artist\u001b[38;5;241m.\u001b[39mget_agg_filter() \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n",
      "File \u001b[0;32m~/miniconda3/envs/bio-compchem/lib/python3.8/site-packages/matplotlib/axis.py:1377\u001b[0m, in \u001b[0;36mAxis.draw\u001b[0;34m(self, renderer, *args, **kwargs)\u001b[0m\n\u001b[1;32m   1374\u001b[0m renderer\u001b[38;5;241m.\u001b[39mopen_group(\u001b[38;5;18m__name__\u001b[39m, gid\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mget_gid())\n\u001b[1;32m   1376\u001b[0m ticks_to_draw \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_update_ticks()\n\u001b[0;32m-> 1377\u001b[0m tlb1, tlb2 \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_get_ticklabel_bboxes\u001b[49m\u001b[43m(\u001b[49m\u001b[43mticks_to_draw\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mrenderer\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1379\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m tick \u001b[38;5;129;01min\u001b[39;00m ticks_to_draw:\n\u001b[1;32m   1380\u001b[0m     tick\u001b[38;5;241m.\u001b[39mdraw(renderer)\n",
      "File \u001b[0;32m~/miniconda3/envs/bio-compchem/lib/python3.8/site-packages/matplotlib/axis.py:1304\u001b[0m, in \u001b[0;36mAxis._get_ticklabel_bboxes\u001b[0;34m(self, ticks, renderer)\u001b[0m\n\u001b[1;32m   1302\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m renderer \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m   1303\u001b[0m     renderer \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mfigure\u001b[38;5;241m.\u001b[39m_get_renderer()\n\u001b[0;32m-> 1304\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m ([tick\u001b[38;5;241m.\u001b[39mlabel1\u001b[38;5;241m.\u001b[39mget_window_extent(renderer)\n\u001b[1;32m   1305\u001b[0m          \u001b[38;5;28;01mfor\u001b[39;00m tick \u001b[38;5;129;01min\u001b[39;00m ticks \u001b[38;5;28;01mif\u001b[39;00m tick\u001b[38;5;241m.\u001b[39mlabel1\u001b[38;5;241m.\u001b[39mget_visible()],\n\u001b[1;32m   1306\u001b[0m         [tick\u001b[38;5;241m.\u001b[39mlabel2\u001b[38;5;241m.\u001b[39mget_window_extent(renderer)\n\u001b[1;32m   1307\u001b[0m          \u001b[38;5;28;01mfor\u001b[39;00m tick \u001b[38;5;129;01min\u001b[39;00m ticks \u001b[38;5;28;01mif\u001b[39;00m tick\u001b[38;5;241m.\u001b[39mlabel2\u001b[38;5;241m.\u001b[39mget_visible()])\n",
      "File \u001b[0;32m~/miniconda3/envs/bio-compchem/lib/python3.8/site-packages/matplotlib/axis.py:1304\u001b[0m, in \u001b[0;36m<listcomp>\u001b[0;34m(.0)\u001b[0m\n\u001b[1;32m   1302\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m renderer \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m   1303\u001b[0m     renderer \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mfigure\u001b[38;5;241m.\u001b[39m_get_renderer()\n\u001b[0;32m-> 1304\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m ([\u001b[43mtick\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mlabel1\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_window_extent\u001b[49m\u001b[43m(\u001b[49m\u001b[43mrenderer\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1305\u001b[0m          \u001b[38;5;28;01mfor\u001b[39;00m tick \u001b[38;5;129;01min\u001b[39;00m ticks \u001b[38;5;28;01mif\u001b[39;00m tick\u001b[38;5;241m.\u001b[39mlabel1\u001b[38;5;241m.\u001b[39mget_visible()],\n\u001b[1;32m   1306\u001b[0m         [tick\u001b[38;5;241m.\u001b[39mlabel2\u001b[38;5;241m.\u001b[39mget_window_extent(renderer)\n\u001b[1;32m   1307\u001b[0m          \u001b[38;5;28;01mfor\u001b[39;00m tick \u001b[38;5;129;01min\u001b[39;00m ticks \u001b[38;5;28;01mif\u001b[39;00m tick\u001b[38;5;241m.\u001b[39mlabel2\u001b[38;5;241m.\u001b[39mget_visible()])\n",
      "File \u001b[0;32m~/miniconda3/envs/bio-compchem/lib/python3.8/site-packages/matplotlib/text.py:959\u001b[0m, in \u001b[0;36mText.get_window_extent\u001b[0;34m(self, renderer, dpi)\u001b[0m\n\u001b[1;32m    954\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mRuntimeError\u001b[39;00m(\n\u001b[1;32m    955\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mCannot get window extent of text w/o renderer. You likely \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    956\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mwant to call \u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mfigure.draw_without_rendering()\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m first.\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m    958\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m cbook\u001b[38;5;241m.\u001b[39m_setattr_cm(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mfigure, dpi\u001b[38;5;241m=\u001b[39mdpi):\n\u001b[0;32m--> 959\u001b[0m     bbox, info, descent \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_get_layout\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_renderer\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    960\u001b[0m     x, y \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mget_unitless_position()\n\u001b[1;32m    961\u001b[0m     x, y \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mget_transform()\u001b[38;5;241m.\u001b[39mtransform((x, y))\n",
      "File \u001b[0;32m~/miniconda3/envs/bio-compchem/lib/python3.8/site-packages/matplotlib/text.py:386\u001b[0m, in \u001b[0;36mText._get_layout\u001b[0;34m(self, renderer)\u001b[0m\n\u001b[1;32m    384\u001b[0m clean_line, ismath \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_preprocess_math(line)\n\u001b[1;32m    385\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m clean_line:\n\u001b[0;32m--> 386\u001b[0m     w, h, d \u001b[38;5;241m=\u001b[39m \u001b[43m_get_text_metrics_with_cache\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    387\u001b[0m \u001b[43m        \u001b[49m\u001b[43mrenderer\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mclean_line\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_fontproperties\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    388\u001b[0m \u001b[43m        \u001b[49m\u001b[43mismath\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mismath\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdpi\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mfigure\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mdpi\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    389\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    390\u001b[0m     w \u001b[38;5;241m=\u001b[39m h \u001b[38;5;241m=\u001b[39m d \u001b[38;5;241m=\u001b[39m \u001b[38;5;241m0\u001b[39m\n",
      "File \u001b[0;32m~/miniconda3/envs/bio-compchem/lib/python3.8/site-packages/matplotlib/text.py:97\u001b[0m, in \u001b[0;36m_get_text_metrics_with_cache\u001b[0;34m(renderer, text, fontprop, ismath, dpi)\u001b[0m\n\u001b[1;32m     94\u001b[0m \u001b[38;5;124;03m\"\"\"Call ``renderer.get_text_width_height_descent``, caching the results.\"\"\"\u001b[39;00m\n\u001b[1;32m     95\u001b[0m \u001b[38;5;66;03m# Cached based on a copy of fontprop so that later in-place mutations of\u001b[39;00m\n\u001b[1;32m     96\u001b[0m \u001b[38;5;66;03m# the passed-in argument do not mess up the cache.\u001b[39;00m\n\u001b[0;32m---> 97\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43m_get_text_metrics_with_cache_impl\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m     98\u001b[0m \u001b[43m    \u001b[49m\u001b[43mweakref\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mref\u001b[49m\u001b[43m(\u001b[49m\u001b[43mrenderer\u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mtext\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mfontprop\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcopy\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mismath\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdpi\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/envs/bio-compchem/lib/python3.8/site-packages/matplotlib/text.py:105\u001b[0m, in \u001b[0;36m_get_text_metrics_with_cache_impl\u001b[0;34m(renderer_ref, text, fontprop, ismath, dpi)\u001b[0m\n\u001b[1;32m    101\u001b[0m \u001b[38;5;129m@functools\u001b[39m\u001b[38;5;241m.\u001b[39mlru_cache(\u001b[38;5;241m4096\u001b[39m)\n\u001b[1;32m    102\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m_get_text_metrics_with_cache_impl\u001b[39m(\n\u001b[1;32m    103\u001b[0m         renderer_ref, text, fontprop, ismath, dpi):\n\u001b[1;32m    104\u001b[0m     \u001b[38;5;66;03m# dpi is unused, but participates in cache invalidation (via the renderer).\u001b[39;00m\n\u001b[0;32m--> 105\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mrenderer_ref\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_text_width_height_descent\u001b[49m\u001b[43m(\u001b[49m\u001b[43mtext\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mfontprop\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mismath\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/envs/bio-compchem/lib/python3.8/site-packages/matplotlib/backends/backend_agg.py:234\u001b[0m, in \u001b[0;36mRendererAgg.get_text_width_height_descent\u001b[0;34m(self, s, prop, ismath)\u001b[0m\n\u001b[1;32m    231\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m width, height, descent\n\u001b[1;32m    233\u001b[0m font \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_prepare_font(prop)\n\u001b[0;32m--> 234\u001b[0m \u001b[43mfont\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mset_text\u001b[49m\u001b[43m(\u001b[49m\u001b[43ms\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m0.0\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mflags\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mget_hinting_flag\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    235\u001b[0m w, h \u001b[38;5;241m=\u001b[39m font\u001b[38;5;241m.\u001b[39mget_width_height()  \u001b[38;5;66;03m# width and height of unrotated string\u001b[39;00m\n\u001b[1;32m    236\u001b[0m d \u001b[38;5;241m=\u001b[39m font\u001b[38;5;241m.\u001b[39mget_descent()\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "plt.figure(figsize=(12, 6))\n",
    "plt.suptitle('Distribution of Family Sizes', fontsize=18, y=0.95)\n",
    "colors = ['tab:blue', 'tab:orange', 'tab:green']\n",
    "\n",
    "for n, (dataset_name, dataset) in enumerate(all_dataset.items()):\n",
    "    # Create the subplot\n",
    "    ax = plt.subplot(1, 3, n + 1)\n",
    "    ax.set_title(dataset_name)\n",
    "    ax.set_xlabel(\"Family Size\")\n",
    "    ax.set_ylabel(\"Number of Families\")\n",
    "    \n",
    "    # Plot data\n",
    "    hist, bins, _ = ax.hist(dataset['family_id'], bins=100, color=colors[n])\n",
    "    \n",
    "    # Calculate the threshold for magnification\n",
    "    threshold = int(len(bins) / 2)\n",
    "    \n",
    "    # Create inset plot for magnified view\n",
    "    ax_inset = inset_axes(ax, width=\"40%\", height=\"40%\", loc='upper right')\n",
    "    ax_inset.hist(dataset['family_id'], bins=bins[threshold:], color=colors[n])\n",
    "    ax_inset.set_title('Zoomed-in View')\n",
    "    ax_inset.set_xlabel(\"Family Size\")\n",
    "    ax_inset.set_ylabel(\"Number of Families\")\n",
    "    \n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "45cd0dea",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>family_id</th>\n",
       "      <th>sequence_name</th>\n",
       "      <th>family_accession</th>\n",
       "      <th>aligned_sequence</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PGK</td>\n",
       "      <td>PGK_METKA/8-390</td>\n",
       "      <td>PF00162.19</td>\n",
       "      <td>MDDF...E....FENKWVLLR.IDINSTVI...D...............</td>\n",
       "      <td>MDDFEFENKWVLLRIDINSTVIDGKIEDDERIKRHLGTIKELMEHD...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Complex1_30kDa</td>\n",
       "      <td>A5CES3_ORITB/35-156</td>\n",
       "      <td>PF00329.19</td>\n",
       "      <td>.IIN..KE..QLIE.VLKLLKN..................DKE......</td>\n",
       "      <td>IINKEQLIEVLKLLKNDKELKFTILTDVFAADFPDRNKRFEIVYNL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DUF5319</td>\n",
       "      <td>A0A088QVF3_9CORY/7-127</td>\n",
       "      <td>PF17252.2</td>\n",
       "      <td>FFSSMPRDPFEGDPNDPASFLEPD.....E...PFAPLTDQERREV...</td>\n",
       "      <td>FFSSMPRDPFEGDPNDPASFLEPDEPFAPLTDQERREVLEDLAAVR...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>DUF1249</td>\n",
       "      <td>G4QMM8_GLANF/84-197</td>\n",
       "      <td>PF06853.12</td>\n",
       "      <td>..LKLLPDCDS.EDLSYEFKISE..........................</td>\n",
       "      <td>LKLLPDCDSEDLSYEFKISEQLKYNIKILDSARYTSTIMMSQVANN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Methyltransf_25</td>\n",
       "      <td>G8YSU8_PICSO/41-136</td>\n",
       "      <td>PF13649.6</td>\n",
       "      <td>VLDVGCGP.G...S..IT...VD..IA.SR.......VP..K.......</td>\n",
       "      <td>VLDVGCGPGSITVDIASRVPKGHVTGIDTFEELISQGQQKASDLKQ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13645</th>\n",
       "      <td>ILVD_EDD</td>\n",
       "      <td>G3YB96_ASPNA/56-590</td>\n",
       "      <td>PF00920.21</td>\n",
       "      <td>RPIIGIINTGSGFNPC..........HGNTPQ.LIEAAKRGIHLNG...</td>\n",
       "      <td>RPIIGIINTGSGFNPCHGNTPQLIEAAKRGIHLNGGIAIDFPTISL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13646</th>\n",
       "      <td>DUF4191</td>\n",
       "      <td>Q5YZ76_NOCFA/22-240</td>\n",
       "      <td>PF13829.6</td>\n",
       "      <td>KQASKERRQQLWQAFQMQRKEDKL.LLPLMIGALVGITALFLVIGF...</td>\n",
       "      <td>KQASKERRQQLWQAFQMQRKEDKLLLPLMIGALVGITALFLVIGFV...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13647</th>\n",
       "      <td>L6_membrane</td>\n",
       "      <td>F6XC26_MONDO/8-207</td>\n",
       "      <td>PF05805.12</td>\n",
       "      <td>ACSRSCSRILGVSLGLSALFAAGANLALFFPDWEVTS.L.....RN...</td>\n",
       "      <td>ACSRSCSRILGVSLGLSALFAAGANLALFFPDWEVTSLRNGLIGKH...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13648</th>\n",
       "      <td>ART-PolyVal</td>\n",
       "      <td>A0A1I2I1G3_9BURK/841-993</td>\n",
       "      <td>PF18760.1</td>\n",
       "      <td>DGRGEPLVVYRGDA.KG......LSVFNGERHGSRV.......QGS...</td>\n",
       "      <td>DGRGEPLVVYRGDAKGLSVFNGERHGSRVQGSIFFASSPYVARGYT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13649</th>\n",
       "      <td>PMEI</td>\n",
       "      <td>PME18_ARATH/56-203</td>\n",
       "      <td>PF04043.15</td>\n",
       "      <td>...SQIC.HG...........AH..DQ......DS...CQA.LLS....</td>\n",
       "      <td>SQICHGAHDQDSCQALLSEFTTLSLSKLNRLDLLHVFLKNSVWRLE...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1064950 rows  5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             family_id             sequence_name family_accession  \\\n",
       "0                  PGK           PGK_METKA/8-390       PF00162.19   \n",
       "1       Complex1_30kDa       A5CES3_ORITB/35-156       PF00329.19   \n",
       "2              DUF5319    A0A088QVF3_9CORY/7-127        PF17252.2   \n",
       "3              DUF1249       G4QMM8_GLANF/84-197       PF06853.12   \n",
       "4      Methyltransf_25       G8YSU8_PICSO/41-136        PF13649.6   \n",
       "...                ...                       ...              ...   \n",
       "13645         ILVD_EDD       G3YB96_ASPNA/56-590       PF00920.21   \n",
       "13646          DUF4191       Q5YZ76_NOCFA/22-240        PF13829.6   \n",
       "13647      L6_membrane        F6XC26_MONDO/8-207       PF05805.12   \n",
       "13648      ART-PolyVal  A0A1I2I1G3_9BURK/841-993        PF18760.1   \n",
       "13649             PMEI        PME18_ARATH/56-203       PF04043.15   \n",
       "\n",
       "                                        aligned_sequence  \\\n",
       "0      MDDF...E....FENKWVLLR.IDINSTVI...D...............   \n",
       "1      .IIN..KE..QLIE.VLKLLKN..................DKE......   \n",
       "2      FFSSMPRDPFEGDPNDPASFLEPD.....E...PFAPLTDQERREV...   \n",
       "3      ..LKLLPDCDS.EDLSYEFKISE..........................   \n",
       "4      VLDVGCGP.G...S..IT...VD..IA.SR.......VP..K.......   \n",
       "...                                                  ...   \n",
       "13645  RPIIGIINTGSGFNPC..........HGNTPQ.LIEAAKRGIHLNG...   \n",
       "13646  KQASKERRQQLWQAFQMQRKEDKL.LLPLMIGALVGITALFLVIGF...   \n",
       "13647  ACSRSCSRILGVSLGLSALFAAGANLALFFPDWEVTS.L.....RN...   \n",
       "13648  DGRGEPLVVYRGDA.KG......LSVFNGERHGSRV.......QGS...   \n",
       "13649  ...SQIC.HG...........AH..DQ......DS...CQA.LLS....   \n",
       "\n",
       "                                                sequence  \n",
       "0      MDDFEFENKWVLLRIDINSTVIDGKIEDDERIKRHLGTIKELMEHD...  \n",
       "1      IINKEQLIEVLKLLKNDKELKFTILTDVFAADFPDRNKRFEIVYNL...  \n",
       "2      FFSSMPRDPFEGDPNDPASFLEPDEPFAPLTDQERREVLEDLAAVR...  \n",
       "3      LKLLPDCDSEDLSYEFKISEQLKYNIKILDSARYTSTIMMSQVANN...  \n",
       "4      VLDVGCGPGSITVDIASRVPKGHVTGIDTFEELISQGQQKASDLKQ...  \n",
       "...                                                  ...  \n",
       "13645  RPIIGIINTGSGFNPCHGNTPQLIEAAKRGIHLNGGIAIDFPTISL...  \n",
       "13646  KQASKERRQQLWQAFQMQRKEDKLLLPLMIGALVGITALFLVIGFV...  \n",
       "13647  ACSRSCSRILGVSLGLSALFAAGANLALFFPDWEVTSLRNGLIGKH...  \n",
       "13648  DGRGEPLVVYRGDAKGLSVFNGERHGSRVQGSIFFASSPYVARGYT...  \n",
       "13649  SQICHGAHDQDSCQALLSEFTTLSLSKLNRLDLLHVFLKNSVWRLE...  \n",
       "\n",
       "[1064950 rows x 5 columns]"
      ]
     },
     "execution_count": 88,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "="
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2589b654",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
